src/ott/solvers/linear/sinkhorn_lr.py

```diff
@@ -492,24 +513,123 @@ def recompute_couplings(
     f1, f2, g1_old, g2_old, h_old, _, _, _, _, _ = state_inner
     return recompute_couplings(f1, g1_old, c_q, f2, g2_old, c_r, h_old, gamma)
 
+  def dykstra_update_kernel(
+      self,
+      k_q: jnp.ndarray,
+      k_r: jnp.ndarray,
+      k_g: jnp.ndarray,
+      gamma: float,
+      ot_prob: linear_problem.LinearProblem,
+      min_entry_value: float = 1e-6,
+      tolerance: float = 1e-3,
+      min_iter: int = 0,
+      inner_iter: int = 10,
+      max_iter: int = 10000
+  ) -> Tuple[jnp.ndarray, jnp.ndarray, jnp.ndarray]:
+    """Run Dykstra's algorithm."""
+    # shortcuts for problem's definition.
+    rank = self.rank
+    n, m = ot_prob.geom.shape
+    a, b = ot_prob.a, ot_prob.b
+
+    g_old = k_g
+    v1_old, v2_old = jnp.ones(rank), jnp.ones(rank)
+    u1, u2 = jnp.ones(n), jnp.ones(m)
+
+    q_gi, q_gp = jnp.ones(rank), jnp.ones(rank)
+    q_q, q_r = jnp.ones(rank), jnp.ones(rank)
+    err = jnp.inf
+    state_inner = u1, u2, v1_old, v2_old, g_old, q_gi, q_gp, q_q, q_r, err
+    constants = k_q, k_r, k_g, a, b
+
+    def cond_fn(
+        iteration: int, constants: Tuple[jnp.ndarray, ...],
+        state_inner: Tuple[jnp.ndarray, ...]
+    ) -> bool:
+      del iteration, constants
+      *_, err = state_inner
+      return err > tolerance
+
+    def body_fn(
+        iteration: int, constants: Tuple[jnp.ndarray, ...],
+        state_inner: Tuple[jnp.ndarray, ...], compute_error: bool
+    ) -> Tuple[jnp.ndarray, ...]:
+      # TODO(michalk8): in the future, use `NamedTuple`
+      u1, u2, v1_old, v2_old, g_old, q_gi, q_gp, q_q, q_r, err = state_inner
+      k_q, k_r, k_g, a, b = constants
+
+      # First Projection
+      u1 = a / jnp.dot(k_q, v1_old)
```

Comment:
would 
```
supp_a = a > 0
```
stored before the loop, and then
```
u1 = jnp.where(a >0, a / jnp.dot(k_q, v1_old), 0.0)
```

(and same for `u2`) help you get rid of the issues you had with tests when some elements in `a` or `b` are 0?

Reply:
This wouldn't have helped (I think), since we're not dividing by `0`. The issue was just convergence, `epsilon=0.5, gamma=10`, which was high `gamma`, lowered it to `1.0` and locally it passes.

Reply:
This solved the issue!